File: src/server/nlp/trends.ts
import type { EnrichedPost } from './types';
import type { AnalyzeResponse } from '../../shared/types';

const toDateKey = (date: Date): string => date.toISOString().slice(0, 10);

export const buildTrends = (posts: EnrichedPost[]): AnalyzeResponse['trends'] => {
  const buckets = new Map<
    string,
    { post_count: number; negative_count: number; risk_keyword_count: number }
  >();

  for (const post of posts) {
    const date = toDateKey(post.createdAt);
    const bucket = buckets.get(date) ?? {
      post_count: 0,
      negative_count: 0,
      risk_keyword_count: 0,
    };
    bucket.post_count += 1;
    if (post.sentiment === 'negative') {
      bucket.negative_count += 1;
    }
    bucket.risk_keyword_count += post.riskKeywords.length;
    buckets.set(date, bucket);
  }

  const timeline = [...buckets.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([date, stats]) => ({
      date,
      post_count: stats.post_count,
      negative_count: stats.negative_count,
      negative_ratio:
        stats.post_count > 0 ? stats.negative_count / stats.post_count : 0,
      risk_keyword_count: stats.risk_keyword_count,
    }));

  const ratios = timeline.map((t) => t.negative_ratio);
  const avgNegativeRatio =
    ratios.length > 0 ? ratios.reduce((a, b) => a + b, 0) / ratios.length : 0;

  const latest = timeline.at(-1);
  const spikeDetected =
    latest !== undefined &&
    latest.post_count >= 3 &&
    latest.negative_ratio > avgNegativeRatio + 0.25;

  const spikeMessage = spikeDetected
    ? `Negative ratio on ${latest.date} (${Math.round(latest.negative_ratio * 100)}%) exceeds the average (${Math.round(avgNegativeRatio * 100)}%)`
    : 'No significant negative spike detected';

  return {
    timeline,
    spike_detected: spikeDetected,
    spike_message: spikeMessage,
    avg_negative_ratio: avgNegativeRatio,
  };
};

File: backend/trends.py
"""Trend detection over time: volume, negative ratio, risk keywords."""

from collections import defaultdict
from datetime import datetime, timezone
from typing import TypedDict


class TrendPoint(TypedDict):
    date: str
    post_count: int
    negative_count: int
    negative_ratio: float
    risk_keyword_count: int


class TrendSummary(TypedDict):
    timeline: list[TrendPoint]
    spike_detected: bool
    spike_message: str
    avg_negative_ratio: float


def _parse_date(created_utc: float | int | None) -> str:
    if not created_utc:
        return "unknown"
    return datetime.fromtimestamp(float(created_utc), tz=timezone.utc).strftime(
        "%Y-%m-%d"
    )


def analyze_trends(posts: list[dict]) -> TrendSummary:
    by_date: dict[str, list[dict]] = defaultdict(list)
    for post in posts:
        day = _parse_date(post.get("created_utc"))
        by_date[day].append(post)

    timeline: list[TrendPoint] = []
    for day in sorted(by_date.keys()):
        day_posts = by_date[day]
        neg = sum(1 for p in day_posts if p.get("sentiment") == "negative")
        risk = sum(p.get("risk_keyword_count", 0) for p in day_posts)
        count = len(day_posts)
        timeline.append(
            {
                "date": day,
                "post_count": count,
                "negative_count": neg,
                "negative_ratio": round(neg / count, 3) if count else 0.0,
                "risk_keyword_count": risk,
            }
        )

    ratios = [t["negative_ratio"] for t in timeline if t["date"] != "unknown"]
    avg_ratio = sum(ratios) / len(ratios) if ratios else 0.0
    spike_detected = False
    spike_message = "No significant negative sentiment spike detected."
    if timeline:
        latest = [t for t in timeline if t["date"] != "unknown"]
        if latest:
            last = latest[-1]
            if avg_ratio > 0 and last["negative_ratio"] >= avg_ratio * 1.5:
                spike_detected = True
                spike_message = (
                    f"Negative sentiment ratio rose to {last['negative_ratio']:.0%} "
                    f"(average {avg_ratio:.0%}). Monitor closely."
                )

    return {
        "timeline": timeline,
        "spike_detected": spike_detected,
        "spike_message": spike_message,
        "avg_negative_ratio": round(avg_ratio, 3),
    }
